### PR TITLE
Allow multiple options proposals, check decision before vote

### LINKED ISSUE #75: Additional check for shareSignedVote
Screenshot 2022-02-28 at 16 24 22

### PR DESCRIPTION
**Pull request by AugustoL**
Functions propose, _stake and _vote had been copy pasted from GenesisProtocol and change the line that check the decision used, only that line.

Additionally to make sure that this minimal change dont affect legacy proposals the tests now run only with DXDVotingMachine, no GenesisProtocolMachine is used, in the v1.0.1 it will be removed completely form tests.

- fix #75 
- Allow creating proposals with more than 2 decisions.
- Checks the validity of the vote decision before executing signed or shared votes.

### CODE REVIEW

File: contracts/dxvote/DXDVotingMachine.sol [hunk 1]
```diff
@@ -210,6 +213,246 @@ contract DXDVotingMachine is GenesisProtocol {
...
+    /**
+     * @dev register a new proposal with the given parameters. Every proposal has a unique ID which is being
+     * generated by calculating keccak256 of a incremented counter.
+     * @param _paramsHash parameters hash
+     * @param _proposer address
+     * @param _organization address
+     * @param _choicesAmount the total amount of choices for the proposal
+     */
+    function proposeMultipleChoice(
+        uint256,
```
Comment (rossneilson) [inline contracts/dxvote/DXDVotingMachine.sol:278]: What is this param for, I thought it was for number of choices?

Comment (AugustoL) [reply contracts/dxvote/DXDVotingMachine.sol:278]: no idea, the had it in the GenesisProtocol but never used it...

File: test/dxvote/DXDVotingMachine.js [hunk 2]
```diff
@@ -799,6 +799,12 @@ contract("DXDVotingMachine", function (accounts) {
           .voteDecision,
         0
       );
+      await expectRevert(
+        dxdVotingMachine.contract.signalVote(proposalId, 3, 60000, {
+          from: accounts[3],
+        }),
+        "invalid voteDecision"
+      );
```
Comment (rossneilson) [inline test/dxvote/DXDVotingMachine.js:807]: I guess we can make another issue but we should probably test this well

File: contracts/dxvote/DXDVotingMachine.sol (hunk 1 above)
Comment (AugustoL) [reply contracts/dxvote/DXDVotingMachine.sol:278]: Ok, found it, there is a second voting machine in mainnet, https://etherscan.io/address/0x1C18bAd5a3ee4e96611275B13a8ed062B4a13055#code

And in this one it uses that parama as _numOfChoices, in line 135.

Going to make the change to use the first param as _numOfChoices, and also check how is implemented there.

Comment (rossneilson) [reply contracts/dxvote/DXDVotingMachine.sol:278]: Sounds great

File: test/dxvote/DXDVotingMachine.js (hunk 2 above)
Comment (AugustoL) [reply test/dxvote/DXDVotingMachine.js:807]: We have a PR in progress for this #121

**Comment by rossneilson** [comment]
I havent looked into the code for this really at all but I wonder if it makes more sense for holographic consensus to be binary or whether it should also have options?

**Comment by rossneilson** [comment]
And I'm still reading through but presumably this still has a negative vote right, that is always NO and always empty?

**Comment by AugustoL** [comment]
> 

We need to allow multiple options to be used by the wallet "competition" scheme, the minimum amount of options is 2, the current wallet scheme uses the default binary options (propose method), in the next wallet scheme version we will always submit proposals with "reject" option that will do nothing.

**Comment by AugustoL** [comment]
> And I'm still reading through but presumably this still has a negative vote right, that is always NO and always empty?

This depends on how the scheme treats the option, in our case on what we have now and what we have been using we have YES and NO, in the future you will have a dynamic range of options but we will always submit proposals with and extra negative option that would be the reject.

File: contracts/dxvote/DXDVotingMachine.sol [hunk 3]
```diff
@@ -210,6 +213,246 @@ contract DXDVotingMachine is GenesisProtocol {
...
 
+    /**
+     * @dev register a new proposal with the given parameters. Every proposal has a unique ID which is being
+     * generated by calculating keccak256 of a incremented counter.
+     * @param _paramsHash parameters hash
+     * @param _proposer address
+     * @param _organization address
+     */
+    function propose(
+        uint256,
```
Comment (hhamud) [inline contracts/dxvote/DXDVotingMachine.sol:224]: is this the same [thing](https://github.com/DXgovernance/dxdao-contracts/pull/119/files?diff=unified&w=0#r847278573)?

File: contracts/dxvote/DXDVotingMachine.sol [hunk 4]
```diff
@@ -210,6 +213,246 @@ contract DXDVotingMachine is GenesisProtocol {
...
+        //calc dao bounty
+        uint256 daoBounty = parameters[_paramsHash]
+            .daoBountyConst
+            .mul(averagesDownstakesOfBoosted[proposal.organizationId])
+            .div(100);
+        proposal.daoBountyRemain = daoBounty.max(parameters[_paramsHash].minimumDaoBounty);
+        proposals[proposalId] = proposal;
+        proposals[proposalId].stakes[NO] = proposal.daoBountyRemain; //dao downstake on the proposal
+        numOfChoices[proposalId] = _choicesAmount;
+        emit NewProposal(proposalId, organizations[proposal.organizationId], _choicesAmount, _proposer, _paramsHash);
```
Comment (hhamud) [inline contracts/dxvote/DXDVotingMachine.sol:320]: I think we should encapsulate this with an internal `propose` function that can be shared and used in both `propose` and `proposeWithMultipleChoices`

File: contracts/dxvote/DXDVotingMachine.sol (hunk 1 above)
Comment (hhamud) [reply contracts/dxvote/DXDVotingMachine.sol:278]: I was just as confused about this as well

**Comment by rossneilson** [comment]
> This depends on how the scheme treats the option, in our case on what we have now and what we have been using we have YES and NO, in the future you will have a dynamic range of options but we will always submit proposals with and extra negative option that would be the reject.

Awesome, so this implementation will rely on a wallet scheme supporting it, makes sense

File: contracts/dxvote/DXDVotingMachine.sol [hunk 5]
```diff
@@ -210,6 +216,223 @@ contract DXDVotingMachine is GenesisProtocol {
...
+     * @param _proposer address
+     * @param _organization address
+     */
+    function propose(
+        uint256,
+        bytes32 _paramsHash,
+        address _proposer,
+        address _organization
+    ) external returns (bytes32) {
+        return _propose(NUM_OF_CHOICES, _paramsHash, _proposer, _organization);
```
Comment (rossneilson) [inline contracts/dxvote/DXDVotingMachine.sol:232]: Maybe NUM_OF_CHOICES should be DEFAULT_NUM_OF_CHOICES or BINARY_VOTE_NUM_OF_CHOICES

Comment (AugustoL) [reply contracts/dxvote/DXDVotingMachine.sol:232]: The name of the constant is defined in the inherited GenesisProtocol, I would have choose a different one too :grimacing:

**Pull request merged**